### PR TITLE
[docs] Fix broken links in the documentation index

### PR DESCRIPTION
**Pull request by benrimmington**

### CODE REVIEW

**Comment by benrimmington** [comment]
@swift-ci Please smoke test

**Comment by varungandhi-apple** [comment]
While you're cleaning up things, could you also add brief descriptions for the documents under docs/ABI? That would be great. 😅

File: docs/README.md
Comment (varungandhi-apple) [inline docs/README.md:99]: Are these intended for the future, or are you planning to address these TODOs as part of this PR?

Comment (benrimmington) [reply docs/README.md:99]: I'm not even sure how to categorize the ABI documents, let alone summarize them.

(In any case, I won't merge this until after your pull request, to avoid a conflict.)

Comment (varungandhi-apple) [reply docs/README.md:99]: I took a brief look at some of them.
- Mangling.rst and that seems like reference documentation, in that it is expected to be comprehensive by covering edge cases. Similarly GenericSignature.md, KeyPaths.md and RegisterUsage.md also look like reference documentation.
- CallingConvention.rst looks like it should probably go under the Surveys subsection.

I haven't looked at the others.

Could you describe the difficulty you faced in categorizing the documents? I do have brief descriptions at the top of docs/README.md as well as a link to https://documentation.divio.com/#the-documentation-system. If that's not enough, maybe we need a longer description on how to categorize things.

Comment (benrimmington) [reply docs/README.md:99]: The main difficulty I have is knowing if a document is completely outdated.

(I think it would be helpful to move those you've already identified, into the [/docs/archive](https://github.com/apple/swift/tree/master/docs/archive) folder.)

Then I have trouble choosing between:
- *Tutorials* versus *How-To Guides*
- *Explanations* versus *Reference Guides*

(The term *Reference Guides* confuses me, because [TSPL](https://docs.swift.org/swift-book/) has a *Language Guide* and a *Language Reference*.)

I also think we should discuss whether or not subfolders (e.g. [/docs/refactoring](https://github.com/apple/swift/tree/master/docs/refactoring)) are a good idea. If all documents can be organized into subfolders matching your categories, is the documentation index still needed?

Comment (varungandhi-apple) [reply docs/README.md:99]: > The main difficulty I have is knowing if a document is completely outdated

😐 yeah, I'm not sure how we can make this easier, it requires some understanding of the subsystem..

> Tutorials versus How-To Guides

You've already seen some examples of How-To Guides.

A tutorial would have more hand-holding and be more narrow. Ideally, it would even be machine-checkable (i.e. have the tutorial's steps always passing in CI). A How-To guide is more broad, and it leaves some gaps in-between for specifics.

> Explanations versus Reference Guides

1. Completeness/Exhaustiveness: A reference guide is expected to be complete/exhaustive. An explanation is not.
2. Teaching: A reference guide is not meant to teach. It should assume background knowledge as necessary. It may include brief examples, but not a lot. An explanation is intended to teach a topic.

For example, Cppreference is mostly a reference guide, it's difficult to actually learn C++ by just using Cppreference.  OTOH, the new Compiler Architecture document is intended to _describe_ things at a high level without going into the intricacies of each and every detail, so it fits under Explanation.

> I also think we should discuss whether or not subfolders (e.g. /docs/refactoring) are a good idea. If all documents can be organized into subfolders matching your categories, is the documentation index still needed?

The job of the index as I see is:

1. Explain what the different categories are.
2. Provide brief summaries of what contains what. Skimming a dozen documents is more difficult than skimming one index.

I don't think a directory structure can provide this by itself.

When creating the index, I deliberately did not start moving things into subdirectories to avoid messing with people's "muscle memory". That way, the index is a purely additive change. New documents are added into subdirectories (such as the new How-To Guides).

I guess for out-dated documents, it doesn't matter that much, so moving them to the archive directory seems reasonable... I haven't given this too much thought though...

> (The term Reference Guides confuses me, because TSPL has a Language Guide and a Language Reference.)

I don't think the Language Reference serves as an actual language reference, unlike say Cppreference. 😅  Similarly, the Language Guide dives into too much reference territory. For example, the [Initialization section](https://docs.swift.org/swift-book/LanguageGuide/Initialization.html) goes into too many details, which are arguably more useful for compiler authors than language users.

The [Racket Guide](https://docs.racket-lang.org/guide/) and [Racket Reference](https://docs.racket-lang.org/reference/) are much closer to our 'Explanation' vs 'Reference Guide' distinction.

Maybe we should remove the term 'Guide' from 'Reference Guide' and have it just be 'Reference'?

**Comment by benrimmington** [comment]
@swift-ci Please smoke test

File: docs/README.md
Comment (varungandhi-apple) [inline docs/README.md:178]: I think it would be helpful if the description here (and the other ones below) briefly summarized the contents of the document. For example, say something like:

> Describes what generic signatures are and how they are used in the ABI, including the algorithm used for minimization and canonicalization.

Nit: Please start the description with a verb (describes/provides/outlines/summarizes/explains etc.), otherwise it reads a bit odd IMO.

File: docs/README.md
Comment (benrimmington) [reply docs/README.md]: > New documents are added into subdirectories (such as the new How-To Guides).

I'm not sure if subdirectories are a good idea, except for the [archive](https://github.com/apple/swift/tree/master/docs/archive) and [proposals](https://github.com/apple/swift/tree/master/docs/proposals). A flat list in /docs/ is easier to search, and some documentation (e.g. [WebAssembly.md](https://github.com/apple/swift/blob/master/docs/WebAssembly.md)) might transition between categories as it evolves.

More images could be moved into the [assets](https://github.com/apple/swift/tree/master/docs/assets) subdirectory. Or there could be a naming convention to associate images with their documentation (e.g. "/docs/FirstPullRequest-GitHubCreatePRScreenshot.png").

The /docs/README.md file could be used as the start page of <https://apple.github.io/swift/>, if [GitHub Pages](https://docs.github.com/pages/) is enabled for this repository. I don't know if `.rst` files are supported, but `.md` files (with YAML front matter) will be converted into `.html` files.

**Comment by benrimmington** [comment]
@swift-ci Please smoke test

**Comment by benrimmington** [comment]
@swift-ci Please smoke test

**Comment by benrimmington** [comment]
@swift-ci Please smoke test macOS platform

**Comment by benrimmington** [comment]
@swift-ci Please clean smoke test macOS platform

**Comment by benrimmington** [comment]
@swift-ci Please smoke test macOS platform

**Comment by benrimmington** [comment]
@swift-ci Please smoke test macOS platform

**Review by varungandhi-apple — APPROVED** [review]
Thank you for addressing the comments!

**Comment by benrimmington** [comment]
@swift-ci Please clean smoke test macOS platform

**Pull request merged**
